File: leetcode/java/0033. Search in Rotated Sorted Array.java
package leetcode.java;

// Time complexity: O(logN)
// Space complexity: O(1)

class Solution {
    public int search(int[] nums, int target) {
        int pivot = findPivot(nums);
        if (nums[pivot] == target) {
            return pivot;
        }

        int leftSearch = bsearch(0, pivot-1, nums, target);
        return leftSearch == -1 ? bsearch(pivot+1, nums.length-1, nums, target) : leftSearch;
    }

    public int bsearch(int left, int right, int[] nums, int target) {
        
        while (right >= left) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }

    public int findPivot(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        
        if (nums.length == 1) {
            return 0;
        }

        while (right >= left) {
            int mid = (left + right) / 2;
            
            if (nums[mid] < nums[mid-1 == -1 ? right : mid-1]) {
                return mid;
            } else if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return 0;
    }
}

File: leetcode/python/0033. Search in Rotated Sorted Array.py
# Time complexity: O(logN)
# Space complexity: O(1)

class Solution(object):
    def search(self, nums, target):
        pivot = self.find_pivot(nums)
        firsthalf = self.bsearch(nums, target, 0, pivot-1)
        if firsthalf == -1:
            return self.bsearch(nums, target, pivot, len(nums)-1)
        return firsthalf


    def find_pivot(self, nums):
        left = 0
        right = len(nums) - 1
        if len(nums) == 1:
            return 0

        while right >= left:
            mid = (left + right) / 2

            if nums[mid] < nums[mid-1]:
                return mid
            elif nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid - 1
        return -1

    def bsearch(self, nums, target, left, right):
        while right >= left:
            mid = (left + right) / 2

            if nums[mid] == target:
                return mid
            elif target > nums[mid]:
                left = mid + 1
            else:
                right = mid - 1
        return -1